python/dgl/nn/pytorch/conv/grouprevres.py

```diff
@@ -0,0 +1,186 @@
+
+        # compute gradients
+        with torch.set_grad_enabled(True):
+            detached_inputs = []
+            for i, element in enumerate(inputs):
+                if isinstance(element, torch.Tensor):
+                    element = element.detach()
+                    element.requires_grad = ctx.input_requires_grad[i]
+                detached_inputs.append(element)
+
+            detached_inputs = tuple(detached_inputs)
+            temp_output = ctx.fn(*detached_inputs)
+
+        filtered_detached_inputs = tuple(filter(lambda x: x.requires_grad, detached_inputs))
+        gradients = torch.autograd.grad(outputs=(temp_output,),
+                                        inputs=filtered_detached_inputs + ctx.weights,
+                                        grad_outputs=grad_outputs)
+
+        input_gradients = []
+        i = 0
+        for rg in ctx.input_requires_grad:
+            if rg:
+                input_gradients.append(gradients[i])
+                i += 1
+            else:
+                input_gradients.append(None)
+
+        gradients = tuple(input_gradients) + gradients[-len(ctx.weights):]
+
+        return (None, None, None) + gradients
+
+
+class GroupRevRes(nn.Module):
+    r"""Grouped reversible residual connections for GNNs, as introduced in
+    `Training Graph Neural Networks with 1000 Layers <https://arxiv.org/abs/2106.07476>`__
+
+    Parameters
+    ----------
+    gnn_module : nn.Module
+        GNN module for message passing. The input and output node representation size need
+        to be the same.
+    group : int, optional
+        The input node features will be partitioned into the specified number of groups.
+
+    Examples
+    --------
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl.nn import GraphConv, GroupRevRes
+
+    >>> num_nodes = 5
+    >>> num_edges = 20
+    >>> feats = 32
+    >>> group = 2
+    >>> g = dgl.rand_graph(num_nodes, num_edges)
+    >>> h = torch.randn(num_nodes, feats)
+    >>> conv = GraphConv(feats // group, feats // group)
+    >>> model = GroupRevRes(conv, group)
+    >>> out = model(g, h)
+    """
+    def __init__(self, gnn_module, group=2):
+        super(GroupRevRes, self).__init__()
+        self.gnn_modules = nn.ModuleList()
+        for i in range(group):
+            if i == 0:
+                self.gnn_modules.append(gnn_module)
+            else:
+                self.gnn_modules.append(deepcopy(gnn_module))
+        self.group = group
+
+    def _forward(self, x, g, *args):
+        xs = torch.chunk(x, self.group, dim=-1)
+
+        if len(args) == 0:
+            args_chunks = [()] * self.group
+        else:
+            chunked_args = list(map(lambda arg: torch.chunk(arg, self.group, dim=-1), args))
+            args_chunks = list(zip(*chunked_args))
+        y_in = sum(xs[1:])
+
+        ys = []
+        for i in range(self.group):
+            tmp_out = self.gnn_modules[i](g, y_in, *args_chunks[i])
+            y = xs[i] + tmp_out
```

Comment:
Consider refactor it to the following to make it close to the math in paper:
```python
ys = [None] * len(xs)
ys[0] = sum(xs[1:])
for i in range(1, self.group):
    ys[i] = self.gnn_modules[i](g, ys[i-1]) + xs[i]
```

Reply:
Strictly speaking, if you really want that, you need to have

```python
ys = [None] * (len(xs) + 1)
ys[0] = sum(xs[1:])
for i in range(self.group):
    ys[i + 1] = self.gnn_modules[i](g, ys[i]) + xs[i]
```